internal/cephfs/clone.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2020 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cephfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	klog "k8s.io/klog/v2"
+)
+
+const (
+	// cephFSCloneFailed indicates that clone is in failed state.
+	cephFSCloneFailed = "failed"
+	// cephFSCloneCompleted indicates that clone is in in-progress state.
+	cephFSCloneInprogress = "in-progress"
+	// cephFSCloneComplete indicates that clone is in complete state.
+	cephFSCloneComplete = "complete"
+)
+
+func createCloneFromSubvolume(ctx context.Context, volID, cloneID volumeID, volOpt, parentvolOpt *volumeOptions, cr *util.Credentials) error {
+	snapshotID := cloneID
+	err := createSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s %v"), snapshotID, err)
+		return err
+	}
+	var (
+		// if protectErr is not nil we will delete the snapshot as the protect fails
+		protectErr error
+		// if cloneErr is not nil we will unprotect the snapshot and delete the snapshot
+		cloneErr error
+	)
+	defer func() {
+		if protectErr != nil {
+			err = deleteSnapshot(ctx, parentvolOpt, cr, snapshotID, volID)
+			if err != nil {
+				klog.Errorf(util.Log(ctx, "failed to delete snapshot %s %v"), snapshotID, err)
+			}
+		}
+
+		if cloneErr != nil {
```

Comment:
Reverse the order of cleanup here to be safe,
- purge clone first
- then unprotect and delete snapshot

IOW, reverse of create workflow.

(also, remove TODO at line 87, as we use `cloneErr` now to cleanup the remanants)

Reply:
Done.. thanks .

Reply:
@ShyamsundarR